src/lib/replication/repl_dev/raft_state_machine.cpp

```diff
@@ -0,0 +1,510 @@
+#include <iomgr/iomgr_timer.hpp>
+#include <sisl/logging/logging.h>
+#include <sisl/fds/utils.hpp>
+#include <sisl/fds/vector_pool.hpp>
+
+#include "repl_dev/raft_state_machine.h"
+#include "repl_dev/raft_repl_dev.h"
+
+SISL_LOGGING_DECL(replication)
+
+namespace homestore {
+
+RaftStateMachine::RaftStateMachine(RaftReplDev& rd) : m_rd{rd}, m_group_id{m_rd.group_id()} {
+    m_success_ptr = nuraft::buffer::alloc(sizeof(int));
+    m_success_ptr->put(0);
+}
+
+raft_buf_ptr_t RaftStateMachine::pre_commit_ext(nuraft::state_machine::ext_op_params const& params) {
+    // Leader precommit is processed in next callback, since lsn would not have been known to repl layer till we get
+    // the next callback.
+    if (!m_rd.is_leader()) {
+        int64_t lsn = s_cast< int64_t >(params.log_idx);
+        raft_buf_ptr_t data = params.data;
+
+        RD_LOG(DEBUG, "pre_commit: {}, size: {}", lsn, data->size());
+        repl_req_ptr_t rreq = lsn_to_req(lsn);
+
+        m_rd.m_listener->on_pre_commit(rreq->lsn, rreq->header, rreq->key, rreq);
+    }
+    return m_success_ptr;
+}
+
+void RaftStateMachine::after_precommit_in_leader(nuraft::raft_server::req_ext_cb_params const& params) {
+    repl_req_ptr_t rreq = repl_req_ptr_t(r_cast< repl_req_ctx* >(params.context));
```

Comment:
log is stored before `pre_commit_ext`

https://github.com/eBay/NuRaft/blob/ff3d2c7a8d73754d3c09dc4b26b7968181f5bcf6/src/handle_client_request.cxx#L135
`params.context` seems not changed between `pre_commit_ext` and `after_pre_commit`, so lsn should be known in `pre_commit_ext`?

Reply:
As explained on Xioaxi's comments, the purpose is our ability to stick the context (which is req structure) on callback, which is not possible with pre_commit_ext. That is why we have another callback to put the context.